domain/consensus/consensus.go

```diff
@@ -881,3 +886,62 @@ func (s *consensus) IsChainBlock(blockHash *externalapi.DomainHash) (bool, error
 
 	return s.dagTopologyManagers[0].IsInSelectedParentChainOf(stagingArea, blockHash, virtualGHOSTDAGData.SelectedParent())
 }
+
+func (s *consensus) HighestChainBlockInPastOf(blockHash *externalapi.DomainHash) (*externalapi.DomainHash, bool, error) {
+	s.lock.Lock()
+	defer s.lock.Unlock()
+
+	stagingArea := model.NewStagingArea()
+	pruningPoint, err := s.pruningStore.PruningPoint(s.databaseContext, stagingArea)
+	if err != nil {
+		return nil, false, err
+	}
+
+	isPruningPointAncestorOfBlock, err := s.dagTopologyManagers[0].IsAncestorOf(stagingArea, pruningPoint, blockHash)
+	if err != nil {
+		return nil, false, err
+	}
+
+	if !isPruningPointAncestorOfBlock {
+		return nil, false, nil
+	}
+
+	headersSelectedTip, err := s.headersSelectedTipStore.HeadersSelectedTip(s.databaseContext, stagingArea)
+	if err != nil {
+		return nil, false, err
+	}
+
+	lowHash := pruningPoint
+	highHash := headersSelectedTip
+	for {
+		blockLocator, err := s.syncManager.CreateHeadersSelectedChainBlockLocator(stagingArea, lowHash, highHash)
```

Comment:
The code might be simpler if we directly implement a binary search here w/o going through block locator logic (also makes it `log(n)` over `(log(n))^2`). Though I can see why reusing has benefits.